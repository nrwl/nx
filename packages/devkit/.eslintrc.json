{
  "extends": "../../.eslintrc.json",
  "ignorePatterns": ["!**/*", "node_modules"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {
        "@typescript-eslint/explicit-module-boundary-types": "error",
        "@typescript-eslint/no-restricted-imports": [
          "error",
          {
            "paths": [
              "@nx/workspace",
              "@angular-devkit/core",
              "@angular-devkit/architect",
              "@angular-devkit/schematics"
            ],
            "patterns": [
              {
                "group": [
                  "nx/bin/*",
                  "nx/src/*",
                  "nx/plugins/*",
                  "!nx/src/devkit-internals",
                  "!nx/src/devkit-exports"
                ],
                "allowTypeImports": true,
                "message": "Only import from nx/src/devkit-internals or nx/src/devkit-exports"
              },
              {
                "group": ["@nx/devkit/**/*"],
                "message": "Use a relative import"
              }
            ]
          }
        ]
      }
    },
    {
      "files": ["*.spec.ts"],
      "rules": {
        "@typescript-eslint/no-restricted-imports": "off"
      }
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["./package.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/nx-plugin-checks": "error"
      }
    },
    {
      "files": ["./package.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/dependency-checks": [
          "error",
          {
            "buildTargets": ["build-base"],
            "ignoredDependencies": [
              "nx",
              "typescript",
              "prettier",
              "tslib",
              // Installed to workspace by plugins
              "rxjs",
              "@angular-devkit/core",
              "@angular-devkit/architect",
              "@angular-devkit/schematics",
              "webpack"
            ]
          }
        ]
      }
    }
  ]
}
