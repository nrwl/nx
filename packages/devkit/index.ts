/**
 * The Nx Devkit is the underlying technology used to customize Nx to support
 * different technologies and custom use-cases. It contains many utility
 * functions for reading and writing files, updating configuration,
 * working with Abstract Syntax Trees(ASTs), and more.
 *
 * As with most things in Nx, the core of Nx Devkit is very simple.
 * It only uses language primitives and immutable objects
 * (the tree being the only exception).
 *
 * @module @nx/devkit
 */

// eslint-disable-next-line @typescript-eslint/no-restricted-imports
export * from 'nx/src/devkit-exports';
export * from './public-api';
