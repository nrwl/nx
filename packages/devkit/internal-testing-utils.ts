/* eslint-disable @nx/enforce-module-boundaries */
export * from '../nx/src/internal-testing-utils/assert-valid-migrations';
export * from '../nx/src/internal-testing-utils/run-migration-against-this-workspace';
export * from '../nx/src/internal-testing-utils/with-environment';
export * from '../nx/src/internal-testing-utils/temp-fs';
/* eslint-enable @nx/enforce-module-boundaries */
export { setCwd } from './src/generators/artifact-name-and-directory-utils';
