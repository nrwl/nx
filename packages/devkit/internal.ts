export { signalToCode } from 'nx/src/devkit-internals';
