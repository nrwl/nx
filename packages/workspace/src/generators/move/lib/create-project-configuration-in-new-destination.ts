import {
  addProjectConfiguration,
  joinPathFragments,
  ProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { NormalizedSchema } from '../schema';

export function createProjectConfigurationInNewDestination(
  tree: Tree,
  schema: NormalizedSchema,
  projectConfig: ProjectConfiguration
) {
  projectConfig.name = schema.newProjectName;
  const isRootProject = projectConfig.root === '.';

  // Subtle bug if project name === path, where the updated name was being overridden.
  const { name, ...rest } = projectConfig;

  // replace old root path with new one
  let newProjectString = JSON.stringify(rest);
  if (isRootProject) {
    // Don't replace . with new root since it'll match all characters.
    // Only look for "./" and replace with new root.
    newProjectString = newProjectString.replace(
      /\.\//g,
      schema.relativeToRootDestination + '/'
    );
    newProjectString = newProjectString.replace(
      /"((tsconfig|jest|webpack|vite)\..*?\.(ts|js|json))"/g,
      `"${schema.relativeToRootDestination}/$1"`
    );
    newProjectString = newProjectString.replace(
      /"(\.\/)?src\/(.*?)"/g,
      `"${schema.relativeToRootDestination}/src/$2"`
    );
  } else {
    // There's another issue if project name === path, where the target
    // string (my-app:build:production) was being replaced
    // and resulting in my-destination/my-new-name:build:production

    // Change anything but target strings (my-app:build:production).
    // Target string are going to be updated in the updateBuildTargets function
    newProjectString = newProjectString.replace(
      new RegExp(projectConfig.root + '(?!:)', 'g'),
      schema.relativeToRootDestination
    );
  }

  const newProject: ProjectConfiguration = {
    name,
    ...JSON.parse(newProjectString),
  };

  newProject.root = schema.relativeToRootDestination;

  // Correct "e2e" target and config since part of the rename will be wrong unless we make the project name "e2e" more unique.
  // e.g. my-app-e2e is safer to search and replace than "e2e".
  if (projectConfig.name === 'e2e') {
    for (const [targetName, targetConfig] of Object.entries(
      newProject.targets
    )) {
      const wrongName = schema.relativeToRootDestination;
      if (targetName !== wrongName) continue;

      if (targetConfig.options?.testingType === wrongName) {
        targetConfig.options.testingType = 'e2e';
      }

      newProject.targets['e2e'] = targetConfig;
      delete newProject.targets[targetName];
    }
  }

  // Original sourceRoot is typically 'src' or 'app', but it could be any folder.
  // Make sure it is updated to be under the new destination.
  if (isRootProject && projectConfig.sourceRoot) {
    newProject.sourceRoot = joinPathFragments(
      schema.relativeToRootDestination,
      projectConfig.sourceRoot
    );
  }

  if (schema.isNxConfiguredInPackageJson) {
    // Update the existing project configuration in the package.json
    updateProjectConfiguration(tree, schema.newProjectName, newProject);
  } else {
    // Create a new project with the root replaced
    addProjectConfiguration(tree, schema.newProjectName, newProject);
  }
}
