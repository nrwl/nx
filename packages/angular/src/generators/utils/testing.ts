import type { Tree } from '@nx/devkit';
import { names, readProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Linter } from '@nx/eslint';
import { UnitTestRunner } from '../../utils/test-runners';
import { applicationGenerator } from '../application/application';
import type { Schema as ApplicationOptions } from '../application/schema';
import { componentGenerator } from '../component/component';
import { host } from '../host/host';
import type { Schema as HostOptions } from '../host/schema';
import { libraryGenerator } from '../library/library';
import type { Schema as LibraryOptions } from '../library/schema';
import { remote } from '../remote/remote';
import type { Schema as RemoteOptions } from '../remote/schema';

export async function generateTestApplication(
  tree: Tree,
  options: ApplicationOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await applicationGenerator(tree, {
    ...options,
  });
}

export async function generateTestHostApplication(
  tree: Tree,
  options: HostOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await host(tree, { ...options });
}

export async function generateTestRemoteApplication(
  tree: Tree,
  options: RemoteOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await remote(tree, { ...options });
}

export async function generateTestLibrary(
  tree: Tree,
  options: LibraryOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await libraryGenerator(tree, {
    ...options,
  });
}

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  tree.write('.gitignore', '');

  await libraryGenerator(tree, {
    directory: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleName: false,
    skipFormat: true,
    unitTestRunner: UnitTestRunner.Jest,
    standalone: false,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    path: `${libName}/src/lib/test-button/test-button`,
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  generateModule(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    path: `${libName}/src/lib/barrel/barrel-button/barrel-button`,
    module: 'barrel',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  generateModule(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    path: `${libName}/src/lib/variable-declare/variable-declare-button/variable-declare-button`,
    module: 'variable-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    path: `${libName}/src/lib/variable-declare/variable-declare-view/variable-declare-view`,
    module: 'variable-declare',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  generateModule(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-button/variable-spread-declare-button`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-view/variable-spread-declare-view`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-anotherview/variable-spread-declare-anotherview`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [
  VariableSpreadDeclareButtonComponent,
  VariableSpreadDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  generateModule(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    path: `${libName}/src/lib/static-member-declarations/cmp1/cmp1`,
    module: 'static-member-declarations',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    path: `${libName}/src/lib/static-member-declarations/cmp2/cmp2`,
    module: 'static-member-declarations',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  generateModule(tree, {
    name: 'nested',
    project: libName,
    path: `${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    module: 'nested',
    path: `${libName}/src/lib/nested/nested-button/nested-button`,
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    path: `${libName}/src/lib/test-other/test-other`,
    standalone: false,
    skipFormat: true,
  });

  return tree;
}

function generateModule(
  tree: Tree,
  options: { name: string; project: string; path?: string }
): void {
  const project = readProjectConfiguration(tree, options.project);

  if (options.path === undefined) {
    const sourceRoot = project.sourceRoot ?? `${project.root}/src`;
    const projectDirName =
      project.projectType === 'application' ? 'app' : 'lib';
    options.path = `${sourceRoot}/${projectDirName}`;
  }

  const moduleNames = names(options.name);
  const moduleFilePath = `${options.path}/${moduleNames.fileName}/${moduleNames.fileName}.module.ts`;

  tree.write(
    moduleFilePath,
    `import { NgModule } from '@angular/core';
  import { CommonModule } from '@angular/common';

  @NgModule({
    declarations: [],
    imports: [CommonModule],
  })
  export class ${moduleNames.className}Module {}`
  );
}
