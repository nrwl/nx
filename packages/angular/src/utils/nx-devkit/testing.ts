import type { Tree } from '@nx/devkit';
import { addProjectConfiguration, names, updateJson } from '@nx/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}

export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}

export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
${routing ? "import { RouterModule } from '@angular/router'" : ''};
import { App } from './app';
@NgModule({
  imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
  declarations: [App],
  bootstrap: [App]
})
export class AppModule {}
`
  );
  tree.write(
    `${appName}/src/main.ts`,
    `import { platformBrowser } from '@angular/platform-browser';

import { AppModule } from './app/app.module';

platformBrowser()
  .bootstrapModule(AppModule)
  .catch(err => console.log(err));
  `
  );
  tree.write(
    `${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  addProjectConfiguration(tree, appName, {
    root: `${appName}`,
    sourceRoot: `${appName}/src`,
    targets: {
      build: {
        options: {
          main: `${appName}/src/main.ts`,
        },
      },
      serve: {
        options: {},
      },
    },
    tags: [],
  });
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
@NgModule({
  imports: [
    CommonModule
  ],
  providers: []
})
export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `export * from './lib/${fileName}.module';
`
  );
}
