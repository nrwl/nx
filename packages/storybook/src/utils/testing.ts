import { Tree } from '@nx/devkit';

import { libraryGenerator } from '@nx/js';
import { Linter } from '@nx/eslint';

export async function createTestUILibNoNgDevkit(
  appTree: Tree,
  libName: string
): Promise<Tree> {
  await libraryGenerator(appTree, {
    linter: Linter.EsLint,
    skipFormat: true,
    skipTsConfig: false,
    unitTestRunner: 'none',
    directory: libName,
  });

  return appTree;
}

export function deleteNewConfigurationAndCreateNew(
  appTree: Tree,
  projectStorybookRoot: string
): Tree {
  // Remove new Storybook configuration
  appTree.delete(`.storybook/main.js`);
  appTree.delete(`.storybook/tsconfig.json`);
  appTree.delete(`${projectStorybookRoot}/main.js`);
  appTree.delete(`${projectStorybookRoot}/preview.js`);
  appTree.delete(`${projectStorybookRoot}/tsconfig.json`);

  // Create old Storybook configuration
  appTree.write(`.storybook/addons.js`, 'console.log("hello")');
  appTree.write(`.storybook/webpack.config.js`, 'console.log("hello")');
  appTree.write(`.storybook/tsconfig.json`, '{"test": "hello"}');
  appTree.write(`${projectStorybookRoot}/config.js`, 'console.log("hello")');
  appTree.write(`${projectStorybookRoot}/addons.js`, 'console.log("hello")');
  appTree.write(
    `${projectStorybookRoot}/webpack.config.js`,
    'console.log("hello")'
  );
  appTree.write(`${projectStorybookRoot}/tsconfig.json`, '{"test": "hello"}');

  return appTree;
}
