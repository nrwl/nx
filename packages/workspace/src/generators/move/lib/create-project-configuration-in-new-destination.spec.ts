import {
  addProjectConfiguration,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import type { NormalizedSchema } from '../schema';
import { createProjectConfigurationInNewDestination } from './create-project-configuration-in-new-destination';

describe('moveProjectConfiguration', () => {
  let tree: Tree;
  let projectConfig: ProjectConfiguration;
  let projectConfigFlat: ProjectConfiguration;
  let schema: NormalizedSchema;

  const setupWorkspace = (version: 1 | 2 = 1) => {
    schema = {
      projectName: 'my-source',
      destination: 'subfolder/my-destination',
      importPath: '@proj/subfolder-my-destination',
      updateImportPath: true,
      newProjectName: 'subfolder-my-destination',
      relativeToRootDestination: 'apps/subfolder/my-destination',
    };

    tree =
      version === 1
        ? createTreeWithEmptyWorkspace()
        : createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    addProjectConfiguration(tree, 'my-source', {
      projectType: 'application',
      root: 'apps/my-source',
      sourceRoot: 'apps/my-source/src',
      targets: {
        build: {
          executor: '@angular-devkit/build-angular:browser',
          options: {
            outputPath: 'dist/apps/my-source',
            index: 'apps/my-source/src/index.html',
            main: 'apps/my-source/src/main.ts',
            polyfills: 'apps/my-source/src/polyfills.ts',
            tsConfig: 'apps/my-source/tsconfig.app.json',
            aot: false,
            assets: [
              'apps/my-source/src/favicon.ico',
              'apps/my-source/src/assets',
            ],
            styles: ['apps/my-source/src/styles.scss'],
            scripts: [],
          },
          configurations: {
            production: {
              fileReplacements: [
                {
                  replace: 'apps/my-source/src/environments/environment.ts',
                  with: 'apps/my-source/src/environments/environment.prod.ts',
                },
              ],
              optimization: true,
              outputHashing: 'all',
              sourceMap: false,
              extractCss: true,
              namedChunks: false,
              aot: true,
              extractLicenses: true,
              vendorChunk: false,
              buildOptimizer: true,
              budgets: [
                {
                  type: 'initial',
                  maximumWarning: '2mb',
                  maximumError: '5mb',
                },
                {
                  type: 'anyComponentStyle',
                  maximumWarning: '6kb',
                  maximumError: '10kb',
                },
              ],
            },
          },
        },
        serve: {
          executor: '@angular-devkit/build-angular:dev-server',
          options: {
            browserTarget: 'my-source:build',
          },
          configurations: {
            production: {
              browserTarget: 'my-source:build:production',
            },
          },
        },
        'extract-i18n': {
          executor: '@angular-devkit/build-angular:extract-i18n',
          options: {
            browserTarget: 'my-source:build',
          },
        },
        lint: {
          executor: '@nx/eslint:lint',
        },
        test: {
          executor: '@nx/jest:jest',
          options: {
            jestConfig: 'apps/my-source/jest.config.js',
            tsConfig: 'apps/my-source/tsconfig.spec.json',
            setupFile: 'apps/my-source/src/test-setup.ts',
          },
        },
      },
      tags: ['type:ui'],
      implicitDependencies: ['my-other-lib'],
    });

    addProjectConfiguration(tree, 'my-source-e2e', {
      root: 'apps/my-source-e2e',
      sourceRoot: 'apps/my-source-e2e/src',
      projectType: 'application',
      targets: {
        e2e: {
          executor: '@nx/cypress:cypress',
          options: {
            cypressConfig: 'apps/my-source-e2e/cypress.json',
            tsConfig: 'apps/my-source-e2e/tsconfig.e2e.json',
            devServerTarget: 'my-source:serve',
          },
          configurations: {
            production: {
              devServerTarget: 'my-source:serve:production',
            },
          },
        },
        lint: {
          executor: '@nx/eslint:lint',
        },
      },
    });

    addProjectConfiguration(tree, 'my-flat-source', {
      root: 'my-flat-source',
      name: 'my-flat-source',
      sourceRoot: 'my-flat-source/src',
      projectType: 'application',
      targets: {
        build: {
          cache: true,
          dependsOn: ['^build'],
          inputs: ['production', '^production'],
          executor: '@nx/vite:build',
          outputs: ['{options.outputPath}'],
          defaultConfiguration: 'production',
          options: { outputPath: 'dist/my-flat-source' },
          configurations: {
            development: { mode: 'development' },
            production: { mode: 'production' },
          },
        },
        serve: {
          executor: '@nx/vite:dev-server',
          defaultConfiguration: 'development',
          options: {
            buildTarget: 'my-flat-source:build:development',
            hmr: true,
          },
          configurations: {
            development: {
              buildTarget: 'my-flat-source:build:development',
              hmr: true,
            },
            production: {
              buildTarget: 'my-flat-source:build:production',
              hmr: false,
            },
          },
        },
        preview: {
          executor: '@nx/vite:preview-server',
          defaultConfiguration: 'development',
          options: { buildTarget: 'my-flat-source:build' },
          configurations: {
            development: { buildTarget: 'my-flat-source:build:development' },
            production: { buildTarget: 'my-flat-source:build:production' },
          },
        },
        test: {
          cache: true,
          inputs: ['default', '^production'],
          executor: '@nx/vite:test',
          outputs: ['{options.reportsDirectory}'],
          options: { reportsDirectory: '../coverage/my-flat-source' },
          configurations: {},
        },
        lint: {
          cache: true,
          inputs: [
            'default',
            '{workspaceRoot}/.eslintrc.json',
            '{workspaceRoot}/.eslintignore',
            '{workspaceRoot}/eslint.config.cjs',
          ],
          executor: '@nx/eslint:lint',
          outputs: ['{options.outputFile}'],
          options: {},
          configurations: {},
        },
      },
      tags: [],
      implicitDependencies: [],
    });

    projectConfig = readProjectConfiguration(tree, 'my-source');
    projectConfigFlat = readProjectConfiguration(tree, 'my-flat-source');
  };

  it('should rename the project', async () => {
    setupWorkspace();

    createProjectConfigurationInNewDestination(tree, schema, projectConfig);

    expect(
      readProjectConfiguration(tree, 'subfolder-my-destination')
    ).toMatchSnapshot();
  });

  it('should rename the project correctly except for build targets', async () => {
    setupWorkspace();

    createProjectConfigurationInNewDestination(
      tree,
      { ...schema, projectName: 'my-flat-source' },
      projectConfigFlat
    );

    const projectConfigFlatDestination = readProjectConfiguration(
      tree,
      'subfolder-my-destination'
    );

    expect(
      projectConfigFlatDestination.targets['preview'].options.buildTarget
    ).toBe('my-flat-source:build');

    expect(projectConfigFlatDestination).toMatchSnapshot();
  });

  it('should update paths in only the intended project', async () => {
    setupWorkspace();

    createProjectConfigurationInNewDestination(tree, schema, projectConfig);

    const actualProject = readProjectConfiguration(
      tree,
      'subfolder-my-destination'
    );
    expect(actualProject).toBeDefined();
    expect(actualProject.root).toBe('apps/subfolder/my-destination');
    expect(actualProject.sourceRoot).toBe('apps/subfolder/my-destination/src');

    const similarProject = readProjectConfiguration(tree, 'my-source-e2e');
    expect(similarProject).toBeDefined();
    expect(similarProject.root).toBe('apps/my-source-e2e');
  });

  it('should update tags and implicitDependencies', () => {
    setupWorkspace();
    createProjectConfigurationInNewDestination(tree, schema, projectConfig);

    const actualProject = readProjectConfiguration(
      tree,
      'subfolder-my-destination'
    );
    expect(actualProject.tags).toEqual(['type:ui']);
    expect(actualProject.implicitDependencies).toEqual(['my-other-lib']);
  });

  it('should support moving a standalone project', () => {
    setupWorkspace(2);

    const projectName = 'standalone';
    const newProjectName = 'parent-standalone';
    addProjectConfiguration(
      tree,
      projectName,
      {
        projectType: 'library',
        root: 'libs/standalone',
        targets: {},
      },
      true
    );
    const moveSchema: NormalizedSchema = {
      projectName: 'standalone',
      destination: 'parent/standalone',
      importPath: '@proj/parent-standalone',
      newProjectName,
      relativeToRootDestination: 'libs/parent/standalone',
      updateImportPath: true,
    };

    createProjectConfigurationInNewDestination(tree, moveSchema, projectConfig);

    expect(readProjectConfiguration(tree, newProjectName)).toBeDefined();
  });
});
