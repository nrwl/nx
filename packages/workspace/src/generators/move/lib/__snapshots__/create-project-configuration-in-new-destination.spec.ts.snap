// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moveProjectConfiguration should rename the project 1`] = `
{
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "implicitDependencies": [
    "my-other-lib",
  ],
  "name": "subfolder-my-destination",
  "projectType": "application",
  "root": "apps/subfolder/my-destination",
  "sourceRoot": "apps/subfolder/my-destination/src",
  "tags": [
    "type:ui",
  ],
  "targets": {
    "build": {
      "configurations": {
        "production": {
          "aot": true,
          "budgets": [
            {
              "maximumError": "5mb",
              "maximumWarning": "2mb",
              "type": "initial",
            },
            {
              "maximumError": "10kb",
              "maximumWarning": "6kb",
              "type": "anyComponentStyle",
            },
          ],
          "buildOptimizer": true,
          "extractCss": true,
          "extractLicenses": true,
          "fileReplacements": [
            {
              "replace": "apps/subfolder/my-destination/src/environments/environment.ts",
              "with": "apps/subfolder/my-destination/src/environments/environment.prod.ts",
            },
          ],
          "namedChunks": false,
          "optimization": true,
          "outputHashing": "all",
          "sourceMap": false,
          "vendorChunk": false,
        },
      },
      "executor": "@angular-devkit/build-angular:browser",
      "options": {
        "aot": false,
        "assets": [
          "apps/subfolder/my-destination/src/favicon.ico",
          "apps/subfolder/my-destination/src/assets",
        ],
        "index": "apps/subfolder/my-destination/src/index.html",
        "main": "apps/subfolder/my-destination/src/main.ts",
        "outputPath": "dist/apps/subfolder/my-destination",
        "polyfills": "apps/subfolder/my-destination/src/polyfills.ts",
        "scripts": [],
        "styles": [
          "apps/subfolder/my-destination/src/styles.scss",
        ],
        "tsConfig": "apps/subfolder/my-destination/tsconfig.app.json",
      },
    },
    "extract-i18n": {
      "executor": "@angular-devkit/build-angular:extract-i18n",
      "options": {
        "browserTarget": "my-source:build",
      },
    },
    "lint": {
      "executor": "@nx/eslint:lint",
    },
    "serve": {
      "configurations": {
        "production": {
          "browserTarget": "my-source:build:production",
        },
      },
      "executor": "@angular-devkit/build-angular:dev-server",
      "options": {
        "browserTarget": "my-source:build",
      },
    },
    "test": {
      "executor": "@nx/jest:jest",
      "options": {
        "jestConfig": "apps/subfolder/my-destination/jest.config.js",
        "setupFile": "apps/subfolder/my-destination/src/test-setup.ts",
        "tsConfig": "apps/subfolder/my-destination/tsconfig.spec.json",
      },
    },
  },
}
`;

exports[`moveProjectConfiguration should rename the project correctly except for build targets 1`] = `
{
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "implicitDependencies": [],
  "name": "subfolder-my-destination",
  "projectType": "application",
  "root": "apps/subfolder/my-destination",
  "sourceRoot": "apps/subfolder/my-destination/src",
  "tags": [],
  "targets": {
    "build": {
      "cache": true,
      "configurations": {
        "development": {
          "mode": "development",
        },
        "production": {
          "mode": "production",
        },
      },
      "defaultConfiguration": "production",
      "dependsOn": [
        "^build",
      ],
      "executor": "@nx/vite:build",
      "inputs": [
        "production",
        "^production",
      ],
      "options": {
        "outputPath": "dist/apps/subfolder/my-destination",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "lint": {
      "cache": true,
      "configurations": {},
      "executor": "@nx/eslint:lint",
      "inputs": [
        "default",
        "{workspaceRoot}/.eslintrc.json",
        "{workspaceRoot}/.eslintignore",
        "{workspaceRoot}/eslint.config.cjs",
      ],
      "options": {},
      "outputs": [
        "{options.outputFile}",
      ],
    },
    "preview": {
      "configurations": {
        "development": {
          "buildTarget": "my-flat-source:build:development",
        },
        "production": {
          "buildTarget": "my-flat-source:build:production",
        },
      },
      "defaultConfiguration": "development",
      "executor": "@nx/vite:preview-server",
      "options": {
        "buildTarget": "my-flat-source:build",
      },
    },
    "serve": {
      "configurations": {
        "development": {
          "buildTarget": "my-flat-source:build:development",
          "hmr": true,
        },
        "production": {
          "buildTarget": "my-flat-source:build:production",
          "hmr": false,
        },
      },
      "defaultConfiguration": "development",
      "executor": "@nx/vite:dev-server",
      "options": {
        "buildTarget": "my-flat-source:build:development",
        "hmr": true,
      },
    },
    "test": {
      "cache": true,
      "configurations": {},
      "executor": "@nx/vite:test",
      "inputs": [
        "default",
        "^production",
      ],
      "options": {
        "reportsDirectory": "../coverage/apps/subfolder/my-destination",
      },
      "outputs": [
        "{options.reportsDirectory}",
      ],
    },
  },
}
`;
